src/org/etools/j1939_84/controllers/part2/Part02Step07Controller.java

```diff
@@ -48,6 +61,102 @@
+                //b. Display each positive return in the log.
+                .reportComponentIdentification(getListener()).getPackets();
+
+        List<OBDModuleInformation> zeroFunctionObdModules = new ArrayList<>();
+
+        final ComponentIdentificationPacket[] dsFunctionZeroPacket = { null };
+        //System.out.println("modules in dataRepo: " + dataRepository.getObdModules().size());
+        //6.2.7 Component ID: Make, Model, Serial Number Support
+        dataRepository.getObdModules().stream()
+                //get modules that responded in Part01 from dataRepo
+                .filter(obdModuleInformation -> obdModuleInformation.getComponentIdentification() != null)
+                .forEach(module -> {
+                    int moduleAddress = module.getSourceAddress();
+                    String moduleName = Lookup.getAddressName(moduleAddress);
+
+                    //6.2.7.1.a Actions: Destination Specific (DS) Component ID request
+                    // (PGN 59904) for PGN 65259 (SPNs 586, 587, and 588) to each OBD ECU.
+                    getVehicleInformationModule()
+                            .reportComponentIdentification(getListener(), moduleAddress)
+                            .getPacket()
+                            .ifPresentOrElse(p -> {
+                                //  6.2.7.2.b Fail criteria: Fail if there is any difference between the
+                                //  part 2 response and the part 1 response, as PGN 65259 data is defined
+                                //  to be static values.
+                                if (p.left.isPresent()) {
+                                    // PGN 65259 -> ComponentIdentificationPacket
+                                    ComponentIdentificationPacket packet = p.left.get();
+                                    if(module.getFunction() == 0){
+                                        dsFunctionZeroPacket[0] = packet;
+                                    }
+                                    if(!packet.getComponentIdentification().equals( module.getComponentIdentification()))
+                                        addFailure("6.2.7.2.b - " + moduleName +
+                                                           " reported component identification as: " + packet.getComponentIdentification().toString() +
+                                                           ", Part 01 Step 09 reported it as: " + module.getComponentIdentification().toString());
+                                    }
+                            }, () -> {
+                                //6.2.7.2 Fail criteria:
+                                // a. Fail if any device does not support PGN 65259 with the engine running that supported PGN 65259 with the engine off in part 1.
+                                addFailure("6.2.7.2.a There are no positive responses to a DS Component ID request from "
```

Comment:
nit: I've been putting a dash "-" between the section number and the error message.